### PR TITLE
feat(installer): expansion-framework support

### PR DESCRIPTION
**Pull request by MaDongdong99**
feat(installer): expansion-framework support

inject expansion framework

start expansion-framework design

expansion-framework design overview

expansion-framework design detailed

expansion-framework design detailed

expansion-framework design detailed

expansion-framework design detailed

expansion-framework design detailed

expansion-framework design detailed

expansion-framework design detailed

expansion-framework design detailed

throw expansion into lib

finish installer expansion inject

split functions into package files

adjust codes

setup application installation of expansion

ignore error at image loaded when non-expansion

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: api/platform/v1/types.go [hunk 1]
```diff
@@ -465,46 +468,46 @@ type Etcd struct {
...
-	CertFile string `json:"certFile" protobuf:"bytes,3,opt,name=certFile"`
+	CertFile string `json:"certFile" protobuf:"bytes,3,opt,name=certFile" yaml:"certFile"`
 
 	// KeyFile is an SSL key file used to secure etcd communication.
 	// Required if using a TLS connection.
-	KeyFile string `json:"keyFile" protobuf:"bytes,4,opt,name=keyFile"`
+	KeyFile string `json:"keyFile" protobuf:"bytes,4,opt,name=keyFile" yaml:"keyFile"`
 }
 
 type Upgrade struct {
```
Comment (huxiaoliang) [inline api/platform/v1/types.go:513]: if you change the type.go, please run `make gen` to generate codes and commit diff as well

File: cmd/tke-installer/app/installer/installer.go [hunk 2]
```diff
@@ -455,6 +483,16 @@ func (t *TKE) initSteps() {
 		}...)
 	}
 
+	// TODO: when expansion enabled, auto set t.Para.Config.Application to be true
```
Comment (huxiaoliang) [inline cmd/tke-installer/app/installer/installer.go:486]: after https://github.com/tkestack/tke/issues/1227,  t.Para.Config.Application is `true` by default

File: api/platform/v1/types.go (hunk 1 above)
Comment (MaDongdong99) [reply api/platform/v1/types.go:513]: done make gen @3cbc9ef

File: cmd/tke-installer/app/installer/chart.go [hunk 3]
```diff
@@ -114,6 +142,38 @@ func (t *TKE) importCharts(ctx context.Context) error {
 				errs = append(errs, err)
 				continue
 			}
+			chartName, _ := parseChartPackage(f)
+			chart := &registryv1.Chart{
```
Comment (huxiaoliang) [inline cmd/tke-installer/app/installer/chart.go:146]: I don't know why need L146-L176， could you please clarify

File: api/application/v1/types.go [hunk 4]
```diff
@@ -55,30 +55,30 @@ type AppList struct {
...
-	RepoPassword   string `json:"repoPassword" protobuf:"bytes,7,opt,name=repoPassword"`
-	ImportedRepo   bool   `json:"importedRepo" protobuf:"bytes,8,opt,name=importedRepo"`
+	TenantID       string `json:"tenantID" protobuf:"bytes,1,opt,name=tenantID" yaml:"tenantID"`
+	ChartGroupName string `json:"chartGroupName" protobuf:"bytes,2,opt,name=chartGroupName" yaml:"chartGroupName"`
+	ChartName      string `json:"chartName" protobuf:"bytes,3,opt,name=chartName" yaml:"chartName"`
+	ChartVersion   string `json:"chartVersion" protobuf:"bytes,4,opt,name=chartVersion" yaml:"chartVersion"`
+	RepoURL        string `json:"repoURL" protobuf:"bytes,5,opt,name=repoURL" yaml:"repoURL"`
+	RepoUsername   string `json:"repoUsername" protobuf:"bytes,6,opt,name=repoUsername" yaml:"repoUsername"`
+	RepoPassword   string `json:"repoPassword" protobuf:"bytes,7,opt,name=repoPassword" yaml:"repoPassword"`
+	ImportedRepo   bool   `json:"importedRepo" protobuf:"bytes,8,opt,name=importedRepo" yaml:"importedRepo"`
```
Comment (leoryu) [inline api/application/v1/types.go:81]: Please don't add yaml tag to types, find some way to transform yaml to json, like k8s.

File: cmd/tke-installer/app/installer/expansion.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	// TODO: transfer tke installer variables to expansion
+	tkeValues := t.values()
+	return t.expansionDriver.InstallApplications(ctx, t.applicationClient, tkeValues)
+}
+
+func (t *TKE) values() map[string]string {
+	ret := make(map[string]string)
+	// TODO: set useful things
+	return ret
+}
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/expansion.go:58]: I don't think these methods are necessary, we should check in next meeting.

File: cmd/tke-installer/app/installer/installer.go [hunk 6]
```diff
@@ -170,6 +177,27 @@ func (t *TKE) loadTKEData() error {
...
+			Func: t.patchHookFiles,
+		},
+	}...)
+	t.steps = append(t.steps, []types.Handler{
+		{
+			Name: "Start Expansion Operator",
+			Func: t.startOperator,
+		},
+	}...)
+
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/installer.go:200]: These are not implemented now, I think we should remove dead code, focus on work which have been implemented.

File: cmd/tke-installer/app/installer/installer.go [hunk 7]
```diff
@@ -455,6 +483,16 @@ func (t *TKE) initSteps() {
...
+	// TODO: when expansion enabled, auto set t.Para.Config.Application to be true
+	if t.Para.Config.Application != nil {
+		t.steps = append(t.steps, []types.Handler{
+			{
+				Name: "Install applications",
+				Func: t.installApplications,
+			},
+		}...)
+	}
+
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/installer.go:495]: May be implemented by applying yaml, not code. This is more common way.

File: cmd/tke-installer/app/installer/installer.go [hunk 8]
```diff
@@ -1001,6 +1039,11 @@ func (t *TKE) findClusterProgress(request *restful.Request, response *restful.Re
 func (t *TKE) do() {
 	ctx := t.log.WithContext(context.Background())
 
+	// Expansion
+	t.mergeExpansionsProviderConf()
+	t.mergeExpansionsConfig()
+	t.mergeExpansionsCluster()
+
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/installer.go:1046]: If we need merge config, we should merge once and forever, no need add new method in future.

**Comment by MaDongdong99** [comment]
instead by  #1269

**Comment by MaDongdong99** [comment]
instead by #1269

**Pull request closed without merging**